Add new components: `neewerlight_ble` and `neewerlight_ct`

khvMX opened the pull request:
# What does this implement/fix?

Adds support for Bi-Color Neewer lights (color temperature & brightness only).

I wanted an esphome component for my Neewer CB60B light.
I have found an RGB one here: https://github.com/litui/esphome-components
The LICENSE is MI, Copyright (c) 2022 Aria Burrell <litui@litui.ca>

I have copied `neewerlight_ble` component verbatim, hence I also added @litui to CODEOWNERS. This component is used to only find out the MAC address.

Then I heavily modified and refactored the `neewerlight` component into `neewerlight_ct`, which now only supports CT lights.

In order to support RGB lights, one would need to first abstract common base as another component (similar to [airthings_wave_base](https://github.com/esphome/esphome/tree/dev/esphome/components/airthings_wave_base)). However, since I have no hardware to test any RGB lights, I have not implemented the abstraction to keep it simple.

Supported Features:
 - turn on/off
 - change light temperature (with configurable range)
 - change brightness

No support for presets, since I manage "presets" a.k.a. Scenes in Home Assistant instead.

Another reference I used for turn/off command and for device name prefixes: https://github.com/taburineagle/NeewerLite-Python
The LICENSE is MIT, Copyright (c) 2021-2022 Zach Glenwright

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- none

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#5078

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

First, find the mac addresses of your Neewer lights:

```yaml
neewerlight_ble:
```

Then, disable it and use `neewerlight_ct` component instead:

```yaml
ble_client:
  - mac_address: AB:CD:EF:00:11:22
    id: neewer_cb60b_1

light:
  - platform: neewerlight_ct
    name: "Neewer CB60B Light"
    ble_client_id: neewer_cb60b_1
    cold_white_color_temperature: 6500 K
    warm_white_color_temperature: 2700 K
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

khvMX committed:
Add neewerlight components from litui to serve as a basis

See https://github.com/litui/esphome-components

khvMX committed:
Refactor NeewerStateOutput out to its own file

khvMX committed:
Refactor color_temp limits to yaml, rename to ct

khvMX committed:
add more prefixes

khvMX committed:
Working implementation for Neewer CB60

khvMX committed:
Add compilation tests (all passed)

khvMX committed:
Change TAGs to neewerlight_ct.component

khvMX committed:
Attempt to fix "sleep" issue

If the light is off (brightness is 0) for some time, when switching it
on again via a single command which sets both brightness and color temp,
the color temperature is set too warm.

The fix idea is to set a different command every time brightness is 0,
namely a "turn on" command, right before actual brightness + color temp
is set.

Not tested yet, have some issues with platformio.

khvMX committed:
Properly handle switch on/off

It seems that the light was never going into standby.
Neewer iphone app did send it to standby when switching off though.
Hopefully this change makes it switch on/off properly
(could not test again)

So the actual problem we were probably facing was:
  - the light has not been switched off, it showed 0% brightness and
    4000K before switching on
  - when switching on it had 7% brightness, but 2700K (presumably, did
    not check)

khvMX committed:
Attempt to fix the bug where the color temperature is set to warmest when the light is turned on.

khvMX committed:
Auto-load esp32_ble_tracker to remove it from yaml

khvMX committed:
Add proper functional tests, but don't push it in PR

khvMX committed:
AUTO_LOAD esp32_ble_tracker

khvMX committed:
Refactor to get rid of extra component

TODO:
  - recheck style guide (at least `this->`)
  - optimize runtime, it got worse

khvMX committed:
Merge branch 'dev' into neewer_lights

khvMX committed:
More cleanup refactoring

khvMX committed:
Add this-> according to C++ code style

See https://developers.esphome.io/contributing/code/

khvMX committed:
Update global CODEOWNERS

khvMX committed:
Remove local tests

khvMX committed:
Merge branch 'dev' into neewer_lights

khvMX committed:
Tiny refactor

khvMX committed:
Fix turning on/off

khvMX committed:
esp32_ble_tracker is already auto-loaded

khvMX committed:
Make linter happy (line endings)

khvMX committed:
Merge branch 'dev' into neewer_lights

khvMX committed:
Fix linter suggestions

khvMX committed:
Few more lint fixes

Lint "error: do not use unnamed namespaces in header files" -> static
Lint "error: unused function 'mireds_to_kelvin_int'" -> ifdef

litui commented:
@khvMX I still have RGB lights if you'd like me to incorporate my original support back in to yours and test. Just let me know.

Also, tagging @DanielBaulig as he expressed some interest in bicolor/CCT support over a year ago in the issues tracker on my repo.

khvMX committed:
Split utils into cpp and h files to fix linting

Static stuff was included in multiple translation units, resulting in
some of the functions not being used. Now, with the proper split, the
functions are only defined once.

khvMX committed:
Merge branch 'dev' into neewer_lights

khvMX commented:
> @khvMX I still have RGB lights if you'd like me to incorporate my original support back in to yours and test. Happy to follow your refactored style if need be, just let me know.

Hey @litui, nice to meet you :) and thanks for the amazing work, I wouldn't be able to do it without it.

Along the way I have fixed 2 little issues which I believe are still in your codebase:
 - normalized color temperature has to be calculated in mireds, your code does it in kelvin, and the values are off (I have a screen on my light, so I could verify what's being set)
 - I have incorporated turn on/off commands, because otherwise it's never truly in standby, and just on zero brightness (again, I could verify by looking at the screen on device, and by hearing the fan not turning off).

> Also, tagging @DanielBaulig as he expressed some interest in bicolor/CCT support over a year ago in the issues tracker on my repo.

Would be great if you could at least test if it works for you @DanielBaulig  ;)

khvMX committed:
Merge branch 'dev' into neewer_lights

khvMX committed:
Act on some of co-pilot suggestions

kbx81 commented on esphome/components/neewerlight_ct/neewer_ct_light_output.cpp:
```diff
@@ -0,0 +1,146 @@
+};
+
+void NeewerCTLightOutput::dump_config() {
+  ESP_LOGCONFIG(TAG, "Neewer CT Light Output:");
+  ESP_LOGCONFIG(TAG, "  MAC address        : %s", this->parent_ ? this->parent_->address_str().c_str() : "N/A");
+  ESP_LOGCONFIG(TAG, "  Service UUID       : %s", this->ble_.service_uuid_.to_string().c_str());
+  ESP_LOGCONFIG(TAG, "  Characteristic UUID: %s", this->ble_.characteristic_uuid_.to_string().c_str());
+  ESP_LOGCONFIG(TAG, "  Require Response   : %s", this->ble_.require_response_ ? "True" : "False");
+  ESP_LOGCONFIG(TAG, "  Colour Temperature Range: %d K - %d K", utils::mireds_to_kelvin_int(warm_white_temperature_),
+                utils::mireds_to_kelvin_int(cold_white_temperature_));
```
Please see https://developers.esphome.io/architecture/logging/#configuration-logging-esp_logconfig

khvMX replied:
Thanks, fixed.

khvMX committed:
Improve dump_config logging

khvMX committed:
Remove extra comma

khvMX committed:
Merge branch 'dev' into neewer_lights

khvMX committed:
Merge branch 'dev' into neewer_lights

khvMX committed:
Merge branch 'dev' into neewer_lights

Commit:
Fix build and lint issues

Commit:
Merge remote-tracking branch 'upstream/dev' into neewer_lights

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

swoboda1337 commented on tests/components/neewerlight_ble/common.yaml:
```diff
@@ -0,0 +1,3 @@
+esp32:
+
```
```suggestion
```

swoboda1337 commented on tests/components/neewerlight_ct/common.yaml:
```diff
@@ -0,0 +1,12 @@
+esp32:
+
```
```suggestion
```

swoboda1337 commented on esphome/components/neewerlight_ct/neewer_state_output.h:
```diff
@@ -0,0 +1,20 @@
+#ifdef USE_ESP32
+
+namespace esphome {
+namespace neewerlight_ct {
```
```suggestion
namespace esphome::neewerlight_ct {
```
Please switch all of them to use the modern style

khvMX replied:
done

swoboda1337 commented on esphome/components/neewerlight_ct/neewer_ble_client.cpp:
```diff
@@ -0,0 +1,75 @@
+    return;
+  }
+  ESP_LOGD(TAG, "[%s] Writing %u bytes", this->characteristic_uuid_.to_string().c_str(), msg.size());
+  for (unsigned int i = 0; i < msg.size(); i++) {
```
Should be using int32_t and uint32_t, please fix all of them

khvMX replied:
changed to `uint32_t`

swoboda1337 commented on esphome/components/neewerlight_ble/__init__.py:
```diff
@@ -0,0 +1,23 @@
+
+def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    yield esp32_ble_tracker.register_ble_device(var, config)
```
Use async/await instead of yield?

khvMX replied:
done

swoboda1337 commented on esphome/components/neewerlight_ble/neewerlight_listener.cpp:
```diff
@@ -0,0 +1,34 @@
+
+static const char *const TAG = "neewerlight_ble";
+
+static const std::unordered_set<std::string> KNOWN_PREFIXES = {
```
Why not just use an array?

khvMX replied:
It's a tool for the job. Why would I use array and keep iterating over it, when I have a hash set? :)

swoboda1337 commented on esphome/components/neewerlight_ct/neewer_ble_client.cpp:
```diff
@@ -0,0 +1,75 @@
+using utils::TAG;
+
+const std::string NeewerBleClient::SERVICE_UUID = "69400001-B5A3-F393-E0A9-E50E24DCCA99";
+const std::string NeewerBleClient::CHARACTERISTIC_UUID = "69400002-B5A3-F393-E0A9-E50E24DCCA99";
```
Better to just do:
```
static constexpr const char *SERVICE_UUID = "69400001-B5A3-F393-E0A9-E50E24DCCA99";
```

khvMX replied:
done. Possibly `static constexpr std::string_view` would have been even better, but not in this particular instance. :)

swoboda1337 commented on esphome/components/neewerlight_ct/neewer_ct_light_output.cpp:
```diff
@@ -0,0 +1,150 @@
+using utils::TAG;
+
+NeewerCTLightOutput::NeewerCTLightOutput() {
+  set_color_temperature(new NeewerStateOutput());
+  set_brightness(new NeewerStateOutput());
```
Missing ->this? Probably best to create these in python looks like a memory leak when its done this way?

khvMX replied:
Hmm, I recall trying that and failing for whatever reason. Indeed, that's a leak.

These are not part of the config, so something like this would be wrong, I presume?

```
    color_temperature = await cg.get_variable(config[CONF_COLOR_TEMPERATURE])
    cg.add(var.set_color_temperature(color_temperature))

    brightness = await cg.get_variable(config[CONF_BRIGHTNESS])
    cg.add(var.set_brightness(brightness))
```

I have to admit, I have little idea how these FloatOutput things work and I don't quite recall the reason for `NeewerStateOutput` with an empty `write_state()`.

Could you please help me sort this out?

swoboda1337 commented on esphome/components/neewerlight_ct/utils.h:
```diff
@@ -0,0 +1,32 @@
+namespace neewerlight_ct {
+namespace utils {
+
+inline constexpr const char TAG[] = "neewerlight_ct.component";
```
inline is redundant for constexpr variables in headers

khvMX replied:
done

swoboda1337 commented on esphome/components/neewerlight_ct/neewer_ble_client.h:
```diff
@@ -0,0 +1,35 @@
+  // NOTE: msg is non-const because BLECharacteristic::write_value failed to declare it as const, just std::move it in
+  void send_message(esphome::ble_client::BLEClient *client, std::vector<uint8_t> &&msg);
+
+  esp32_ble_tracker::ESPBTUUID service_uuid_ = esp32_ble_tracker::ESPBTUUID::from_raw(SERVICE_UUID);
+  esp32_ble_tracker::ESPBTUUID characteristic_uuid_ = esp32_ble_tracker::ESPBTUUID::from_raw(CHARACTERISTIC_UUID);
+  bool require_response_ = true;
```
These should be protected?

khvMX replied:
That was the lazy way to access them in `NeewerCTLightOutput::dump_config()`. Should I move these to protected and implement getters?

swoboda1337 commented on esphome/components/neewerlight_ct/utils.cpp:
```diff
@@ -0,0 +1,53 @@
+}
+
+uint8_t checksum(const std::vector<uint8_t> &data) {
+  uint64_t checksum = 0;
```
```suggestion
  uint8_t checksum = 0;
```

khvMX replied:
done, thanks

swoboda1337 commented on esphome/components/neewerlight_ct/utils.cpp:
```diff
@@ -0,0 +1,53 @@
+    checksum += byte;
+  }
+  uint8_t checksum_byte = checksum & 0xFF;
+  return checksum_byte;
```
```suggestion
  return checksum;
```

khvMX replied:
done

clydebarrow commented on tests/components/neewerlight_ct/test.esp32-c3-ard.yaml:
```diff
@@ -0,0 +1 @@
+<<: !include common.yaml
```
Unless there is conditional compilation for different platforms and chip variants, there's no value in build testing both IDF/Arduino and ESP32/ESP32-C3, and it just makes the CI take longer.

khvMX replied:
removed c3 stuff

clydebarrow commented on esphome/components/neewerlight_ct/utils.h:
```diff
@@ -0,0 +1,32 @@
+#ifdef USE_ESP32
+
+namespace esphome {
+namespace neewerlight_ct {
```
Best to rename this and `message.h` with a prefix like the other files - due to the way the Doxygen API docs are built, each header file should have a globally unique file name.

khvMX replied:
Ok, thanks. done.

clydebarrow commented on esphome/components/neewerlight_ct/neewer_utils.cpp:
```diff
@@ -0,0 +1,53 @@
+
+uint8_t checksum(const std::vector<uint8_t> &data) {
+  uint64_t checksum = 0;
+  for (uint8_t byte : data) {
```
Use `std::accumulate` here.

khvMX replied:
done

clydebarrow commented on esphome/components/neewerlight_ct/neewer_ble_client.cpp:
```diff
@@ -0,0 +1,75 @@
+  switch (event) {
+    case ESP_GATTC_OPEN_EVT:
+      this->client_state_ = esp32_ble_tracker::ClientState::ESTABLISHED;
+      ESP_LOGW(TAG, "[%s] Connected successfully!", this->characteristic_uuid_.to_string().c_str());
```
This would be better as `LOGD` since it's a normal event.

khvMX replied:
done

khvMX committed:
Update tests/components/neewerlight_ble/common.yaml

Co-authored-by: Jonathan Swoboda <154711427+swoboda1337@users.noreply.github.com>

khvMX committed:
Update tests/components/neewerlight_ct/common.yaml

Co-authored-by: Jonathan Swoboda <154711427+swoboda1337@users.noreply.github.com>

Commit:
Fix namespaces

Commit:
Stick to uint32_t

Commit:
yeild -> await

Commit:
Remove `inline`

Commit:
Simplify checksum function

Commit:
Remove redundant tests

"Unless there is conditional compilation for different platforms and
chip variants, there's no value in build testing both IDF/Arduino and
ESP32/ESP32-C3, and it just makes the CI take longer."

Commit:
Rename files for Doxygen compatibility

"Due to the way the Doxygen API docs are built, each header file should have a globally unique file name."

Commit:
Use `std::accumulate`

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

Commit:
`LOGW` -> `LOGD`

Commit:
Merge branch 'neewer_lights' of github.com:khvMX/esphome into neewer_lights

khvMX committed:
Merge branch 'dev' into neewer_lights

Commit:
Fix header inclusion after renaming

Commit:
Add missing include

Commit:
Optimize string constants

khvMX committed:
Merge branch 'dev' into neewer_lights

khvMX committed:
Merge branch 'dev' into neewer_lights
